Fix begin_bit == end_bit == 0 for device-wide and segmented sort

canonizer opened the pull request:
Fix `begin_bit == end_bit == 0` for device-wide and segmented sort.

canonizer commented:
This pull request should address https://github.com/NVIDIA/cub/issues/353.

senior-zero reviewed (changes requested):
@canonizer thank you for addressing this! I'm a bit concerned about the approach, though. I wonder if we could short-circuit in the `begin_bit == end_bit` case. For instance, if `is_overwrite_okay == true` we wouldn't do anything, since the double buffer would contain proper data already. Otherwise, we might just copy the data. I've written a simple benchmark below that might help understand the impact of this approach. For double buffer case, we have noop, which is definitely faster. Otherwise, memcpy is about 40% faster than actually sorting anything.   

```cpp
#include <cub/cub.cuh>

#include <thrust/device_vector.h>

#include <iostream>

void sort(
    std::uint8_t *d_temp_storage, std::size_t &temp_storage_bytes,
    int *d_keys_in, int *d_keys_out, 
    int num_items, 
    bool use_buffer, bool short_circuit) 
{
  const int begin_bit = 0;
  const int end_bit = begin_bit;
  cub::DoubleBuffer<int> d_keys(d_keys_in, d_keys_out);

  if (use_buffer) {
    if (short_circuit) {
      temp_storage_bytes = 1; // noop
    } else {
      cub::DeviceRadixSort::SortKeys(
          d_temp_storage, temp_storage_bytes, 
          d_keys, num_items, begin_bit, end_bit);
    }
  } else {
    if (short_circuit) {
      if (d_temp_storage == nullptr) {
        temp_storage_bytes = 1; 
      } else {
        cudaMemcpy(d_keys_out, d_keys_in, sizeof(int) * num_items, cudaMemcpyDeviceToDevice);
      }
    } else {
      cub::DeviceRadixSort::SortKeys(
          d_temp_storage, temp_storage_bytes, 
          d_keys_in, d_keys_out, num_items, begin_bit, end_bit);
    }
  }
}

int main()
{
  const int num_items = 128 * 1024 * 1024;
  thrust::device_vector<int> in(num_items);
  thrust::device_vector<int> out(num_items);

  int *d_keys_in = thrust::raw_pointer_cast(in.data());
  int *d_keys_out = thrust::raw_pointer_cast(out.data());

  std::uint8_t *d_temp_storage{};
  std::size_t temp_storage_bytes = 0;

  const bool use_buffer = false;
  const bool short_circuit = true;

  sort(d_temp_storage, temp_storage_bytes, d_keys_in, d_keys_out, num_items,
       use_buffer, short_circuit);

  thrust::device_vector<std::uint8_t> temp_storage(temp_storage_bytes);
  d_temp_storage = thrust::raw_pointer_cast(temp_storage.data());

  cudaEvent_t begin, end;
  cudaEventCreate(&begin);
  cudaEventCreate(&end);

  cudaEventRecord(begin);
  sort(d_temp_storage, temp_storage_bytes, d_keys_in, d_keys_out, num_items,
       use_buffer, short_circuit);
  cudaEventRecord(end);
  cudaEventSynchronize(end);

  float ms{};
  cudaEventElapsedTime(&ms, begin, end);

  std::cout << ms << "ms" << std::endl;


  cudaEventDestroy(end);
  cudaEventDestroy(begin);
}
```

I don't think this approach is applicable to segmented version. But I'd like to know your opinion on this for non-segmented API. Are there any downsides I'm missing?

canonizer commented:
@senior-zero @allisonvacanti I've added short-circuiting when `begin_bit == end_bit` and double buffers are passed as arguments. Could you take another look?

allisonvacanti commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1575,15 +1580,11 @@ struct DispatchRadixSort :
-          }
-
-          return cudaSuccess;
+            if (d_temp_storage == nullptr) temp_storage_bytes = 1;
```
Leave the explicit braces / linebreaks in the body here -- we're in the process of migrating code to use the safer style and want the body explicitly bounded.

canonizer replied:
Done.

allisonvacanti commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1925,15 +1926,11 @@ struct DispatchSegmentedRadixSort :
-          }
-
-          return cudaSuccess;
+            if (d_temp_storage == nullptr) temp_storage_bytes = 1;
```
Same as above.

canonizer replied:
Done.

allisonvacanti commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1575,15 +1580,11 @@ struct DispatchRadixSort :
-        // Return if empty problem
-        if (num_items == 0)
+        // Return if empty problem, or if no bits to sort and double-buffering is used
+        if (num_items == 0 || begin_bit == end_bit && is_overwrite_okay)
```
Explicit parens make intent clearer here:

```suggestion
        if (num_items == 0 || (begin_bit == end_bit && is_overwrite_okay))
```

canonizer replied:
Done.

allisonvacanti commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1925,15 +1926,11 @@ struct DispatchSegmentedRadixSort :
-        // Return if empty problem
-        if (num_items == 0)
+        // Return if empty problem, or if no bits to sort and double-buffering is used
+        if (num_items == 0 || begin_bit == end_bit && is_overwrite_okay)
```
```suggestion
        if (num_items == 0 || (begin_bit == end_bit && is_overwrite_okay))
```

canonizer replied:
Done.

allisonvacanti reviewed (changes requested):
In @senior-zero's earlier comment, he suggested making the `is_overwrite_okay == false` case to just do a copy and skip the sorting altogether. Can we add that optimization?

canonizer commented:
Thanks for your comments!

@allisonvacanti I've addressed your comments.

@senior-zero I've added the copy shortcut if `begin_bit == end_bit`, and overwrite is not allowed, for the device-wide sort only. For the segmented sort, as you've mentioned, this approach won't work, and adding a separate kernel to perform segmented copy is definitely for a different pull request.

Could you take another look?

allisonvacanti commented:
@canonizer Can you rebase this on `main`? It looks like there are some conflicts.

senior-zero commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1451,6 +1451,8 @@ struct DispatchRadixSort :
                     }
                 }
 
+                if (CubDebug(error)) break;
```
Is this check needed?

canonizer replied:
The check is necessary to break out of the loop in case of error. An extra `CubDebug()` is not, so I removed it.

senior-zero commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1637,6 +1639,42 @@ struct DispatchRadixSort :
+        
+        // Copy keys
+        cudaError_t error = cudaSuccess;
+        error = cudaMemcpyAsync(d_keys.Alternate(), d_keys.Current(), num_items * sizeof(KeyT),
```
We have to synchronize in the case of `debug_synchronous`
```
if (debug_synchronous)
{
  if (CubDebug(error = SyncStream(stream))
  {
    // ..
```

canonizer replied:
Done (using `DebugSyncStream()`).

canonizer commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1637,6 +1639,42 @@ struct DispatchRadixSort :
+        // Copy values if necessary
+        if (!KEYS_ONLY)
+        {
+            error = cudaMemcpyAsync(d_values.Alternate(), d_values.Current(),
```
Done (using `DebugSyncStream()`).

canonizer commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1637,6 +1639,42 @@ struct DispatchRadixSort :
+        }
+        
+        // Copy keys
+        cudaError_t error = cudaSuccess;
```
Done.

canonizer commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1637,6 +1639,42 @@ struct DispatchRadixSort :
+        // Copy keys
+        cudaError_t error = cudaSuccess;
+        error = cudaMemcpyAsync(d_keys.Alternate(), d_keys.Current(), num_items * sizeof(KeyT),
+                                cudaMemcpyDefault, stream);
```
Replaced with `cudaMempcyDeviceToDevice`.

canonizer commented:
@allisonvacanti @senior-zero I've synced with the latest `main` branch and addressed all review comments. Could you take another look?

senior-zero commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1747,7 +1747,10 @@ struct DispatchRadixSort :
+        cudaError_t error = cudaSuccess;
+        bool has_uva = false;
+        if ((error = HasUVA(has_uva)) != cudaSuccess) return error;
+        if (begin_bit == end_bit & has_uva)
```
This check is not synchronized with the temp storage size estimation above. On systems without uva it'll lead to invalid memory accesses. 

Also, property queries are not free. Can we invoke `HasUVA` only when `begin_bit == end_bit`?

canonizer replied:
Done (for invoking `HasUVA()` only when `begin_bit == end_bit`).

Regarding temporary storage estimate, the one above only applies to the cases where no GPU work at all is performed. If a single copy is required, a separate estimate is provided by `InvokeCopy()` (see the function definition for the actual code).

senior-zero commented on cub/util_device.cuh:
```diff
@@ -125,6 +125,23 @@ CUB_RUNTIME_FUNCTION inline int CurrentDevice()
 }
 
+/** \brief Gets whether the current device supports unified addressing */
+CUB_RUNTIME_FUNCTION cudaError_t HasUVA(bool& has_uva)
```
Can you put this function in the `detail` namespace for now?

canonizer replied:
Done.

canonizer commented:
@senior-zero I've addressed your comments. Could you take another look?

senior-zero commented on cub/device/dispatch/dispatch_radix_sort.cuh:
```diff
@@ -1498,6 +1498,11 @@ struct DispatchRadixSort :
                     }
                 }
 
+                if (error != cudaSuccess)
```
Everything that sets `error` before this place, also checks it, right? I fail to see why this check is needed.

canonizer replied:
This check is needed to break out of the `for (int current_bit = ...)` loop in case of error.

senior-zero replied:
Ah, got it, thanks!

senior-zero reviewed (approved):
Thank you for this optimization! I'll start testing now.

senior-zero commented on cub/util_device.cuh:
```diff
@@ -549,6 +549,23 @@ CUB_RUNTIME_FUNCTION inline cudaError_t DebugSyncStream(cudaStream_t stream)
 }
 
+/** \brief Gets whether the current device supports unified addressing */
+CUB_RUNTIME_FUNCTION cudaError_t HasUVA(bool& has_uva)
```
```suggestion
CUB_RUNTIME_FUNCTION inline cudaError_t HasUVA(bool& has_uva)
```
